File: internal/engine/sqlserver/structure.go
package sqlserver

import (
	"context"
	"strconv"
	"strings"

	"github.com/branow/dbmap/internal/catalog"
	"github.com/branow/dbmap/internal/engine"
)

// source is one metadata query and how to fold its rows into a structure.
// Sources are data so that adding one is a row rather than a branch, and so a
// test can assert what this engine would send with no server behind it.
type source struct {
	Name string
	SQL  func() string
	// Width is the narrowest row this source can still shape. A shorter row is
	// dropped.
	Width int
	// Fold applies one row to the structure it belongs to.
	Fold func(target *parsed, row []string)
}

// parsed is a structure under construction. Index rows arrive one per column,
// so they are held flat until every row is in and then folded.
type parsed struct {
	structure catalog.Structure
	indexRows []indexRow
}

type indexRow struct {
	name    string
	primary bool
	unique  bool
	column  string
}

// sources is the structure stage: five bounded queries, around 8,800 narrow
// metadata rows across the three measured databases. Every one reads sys.*
// catalog views only, so no user data page is touched and none of them needs a
// memory grant worth the name.
var sources = []source{
	{
		Name:  "columns",
		Width: 10,
		SQL: func() string {
			return `SELECT s.name, o.name, c.name, t.name, c.max_length, c.precision, c.scale,
  c.is_nullable, c.is_identity, c.is_computed
FROM sys.columns c
JOIN sys.objects o ON o.object_id = c.object_id
JOIN sys.schemas s ON s.schema_id = o.schema_id
JOIN sys.types t ON t.user_type_id = c.user_type_id
WHERE ` + inScope() + `
ORDER BY s.name, o.name, c.column_id`
		},
		Fold: func(target *parsed, row []string) {
			target.structure.Columns = append(target.structure.Columns, catalog.Column{
				Name:     row[2],
				Type:     row[3],
				Length:   width(row[3], row[4], row[5], row[6]),
				Nullable: flag(row[7]),
				Identity: flag(row[8]),
				Computed: flag(row[9]),
			})
		},
	},
	{
		Name:  "indexes",
		Width: 6,
		SQL: func() string {
			return `SELECT s.name, o.name, i.name, i.is_primary_key, i.is_unique, c.name
FROM sys.indexes i
JOIN sys.objects o ON o.object_id = i.object_id
JOIN sys.schemas s ON s.schema_id = o.schema_id
JOIN sys.index_columns ic ON ic.object_id = i.object_id AND ic.index_id = i.index_id
JOIN sys.columns c ON c.object_id = ic.object_id AND c.column_id = ic.column_id
WHERE ` + inScope() + ` AND i.type > 0 AND ic.is_included_column = 0
ORDER BY s.name, o.name, i.index_id, ic.key_ordinal`
		},
		Fold: func(target *parsed, row []string) {
			target.indexRows = append(target.indexRows, indexRow{
				name:    row[2],
				primary: flag(row[3]),
				unique:  flag(row[4]),
				column:  row[5],
			})
		},
	},
	{
		Name:  "foreignKeys",
		Width: 6,
		SQL: func() string {
			return `SELECT ps.name, po.name, pc.name, rs.name, ro.name, rc.name
FROM sys.foreign_key_columns fkc
JOIN sys.objects po ON po.object_id = fkc.parent_object_id
JOIN sys.schemas ps ON ps.schema_id = po.schema_id
JOIN sys.columns pc ON pc.object_id = fkc.parent_object_id
  AND pc.column_id = fkc.parent_column_id
JOIN sys.objects ro ON ro.object_id = fkc.referenced_object_id
JOIN sys.schemas rs ON rs.schema_id = ro.schema_id
JOIN sys.columns rc ON rc.object_id = fkc.referenced_object_id
  AND rc.column_id = fkc.referenced_column_id
WHERE ps.name NOT IN (` + schemaList() + `)
ORDER BY ps.name, po.name, pc.name`
		},
		Fold: func(target *parsed, row []string) {
			target.structure.ForeignKeys = append(target.structure.ForeignKeys, catalog.ForeignKey{
				Column:     row[2],
				References: row[3] + "." + row[4] + "." + row[5],
			})
		},
	},
	{
		Name:  "parameters",
		Width: 6,
		SQL: func() string {
			return `SELECT s.name, o.name, p.name, t.name, p.max_length, p.is_output
FROM sys.parameters p
JOIN sys.objects o ON o.object_id = p.object_id
JOIN sys.schemas s ON s.schema_id = o.schema_id
JOIN sys.types t ON t.user_type_id = p.user_type_id
WHERE ` + inScope() + `
ORDER BY s.name, o.name, p.parameter_id`
		},
		Fold: func(target *parsed, row []string) {
			name := row[2]
			if name == "" {
				name = catalog.Returns
			}
			target.structure.Parameters = append(target.structure.Parameters, catalog.Param{
				Name:   name,
				Type:   row[3] + width(row[3], row[4], "0", "0"),
				Output: flag(row[5]),
			})
		},
	},
	{
		Name:  "synonyms",
		Width: 3,
		SQL: func() string {
			return `SELECT s.name, sn.name, sn.base_object_name
FROM sys.synonyms sn
JOIN sys.schemas s ON s.schema_id = sn.schema_id
WHERE s.name NOT IN (` + schemaList() + `)
ORDER BY s.name, sn.name`
		},
		Fold: func(target *parsed, row []string) { target.structure.Target = row[2] },
	},
}

// sized are the types whose width a reader needs spelled out.
var sized = []string{"varchar", "nvarchar", "char", "nchar", "varbinary", "binary"}

// scaled are the types that carry a precision and a scale instead.
var scaled = []string{"decimal", "numeric"}

// width renders the width a reader needs — "(50)", "(18,2)", "(max)" — rather
// than the raw catalog numbers. max_length is in BYTES, so a national type
// halves it, and -1 is the (max) sentinel.
func width(kind, maxLength, precision, scale string) string {
	if contains(scaled, kind) {
		return "(" + precision + "," + scale + ")"
	}
	if !contains(sized, kind) {
		return ""
	}
	bytes, err := strconv.Atoi(maxLength)
	if err != nil {
		return ""
	}
	if bytes == -1 {
		return "(max)"
	}
	if strings.HasPrefix(kind, "n") {
		bytes /= 2
	}
	return "(" + strconv.Itoa(bytes) + ")"
}

func contains(list []string, value string) bool {
	for _, item := range list {
		if item == value {
			return true
		}
	}
	return false
}

// fold applies one source's rows to the structures under construction. It is
// pure, so the whole of the structure stage's shaping is testable without a
// connection.
func fold(target map[string]*parsed, source source, rows [][]string) {
	for _, raw := range rows {
		if len(raw) < source.Width {
			continue
		}
		row := cells(raw)
		key := row[0] + "." + row[1]
		entry, ok := target[key]
		if !ok {
			entry = &parsed{}
			target[key] = entry
		}
		source.Fold(entry, row)
	}
}

// foldIndexes turns one-row-per-column index rows into one entry per index and
// lifts the primary key out, because a reader wants it named separately.
func foldIndexes(rows []indexRow) ([]string, []catalog.Index) {
	var order []string
	byName := map[string]*catalog.Index{}
	primary := map[string]bool{}

	for _, row := range rows {
		entry, ok := byName[row.name]
		if !ok {
			entry = &catalog.Index{Name: row.name, Unique: row.unique}
			byName[row.name] = entry
			primary[row.name] = row.primary
			order = append(order, row.name)
		}
		entry.Columns = append(entry.Columns, row.column)
	}

	var key []string
	var indexes []catalog.Index
	for _, name := range order {
		if primary[name] {
			if key == nil {
				key = byName[name].Columns
			}
			continue
		}
		indexes = append(indexes, *byName[name])
	}
	return key, indexes
}

// Structure runs every source against one database, asking for room between
// each, and returns a structure per object key.
func (e *Engine) Structure(
	ctx context.Context,
	conn engine.Conn,
) (map[string]catalog.Structure, error) {
	target := map[string]*parsed{}

	for _, source := range sources {
		if err := e.halt(ctx, conn, "structure/"+source.Name); err != nil {
			return nil, err
		}
		rows, err := e.query(ctx, conn, source.SQL())
		if err != nil {
			return nil, err
		}
		fold(target, source, rows)
	}

	structures := make(map[string]catalog.Structure, len(target))
	for key, entry := range target {
		entry.structure.PrimaryKey, entry.structure.Indexes = foldIndexes(entry.indexRows)
		structures[key] = entry.structure
	}
	return structures, nil
}

File: reference/js/structure.js
"use strict";
// Structure stage: columns, keys, indexes, parameters and synonym targets.
//
// One bounded query per source per database — five round trips, around 8,800
// narrow metadata rows across the three stage databases. Every source reads
// sys.* catalog views only, so no user data page is touched and no query needs
// a memory grant worth the name.
//
// Sources are data. Each declares the SQL it sends and how to shape the rows it
// gets back, so adding one is a row rather than a branch, and a test can assert
// what the build would send without a server.

const scope = require("./scope.js");
const sqlcmd = require("./sqlcmd.js");
const health = require("./health.js");

const typeList = () =>
  scope
    .typeCodes()
    .map((t) => `'${t}'`)
    .join(",");

const schemaList = () => scope.EXCLUDED_SCHEMAS.map((s) => `'${s}'`).join(",");

const IN_SCOPE = () => `o.is_ms_shipped = 0 AND o.type IN (${typeList()}) AND s.name NOT IN (${schemaList()})`;

const SOURCES = [
  {
    name: "columns",
    sql: () => `SELECT s.name, o.name, c.name, t.name, c.max_length, c.precision, c.scale,
  c.is_nullable, c.is_identity, c.is_computed
FROM sys.columns c
JOIN sys.objects o ON o.object_id = c.object_id
JOIN sys.schemas s ON s.schema_id = o.schema_id
JOIN sys.types t ON t.user_type_id = c.user_type_id
WHERE ${IN_SCOPE()}
ORDER BY s.name, o.name, c.column_id`,
    shape: (r) => ({
      owner: `${r[0]}.${r[1]}`,
      value: {
        name: r[2],
        type: r[3],
        length: widthOf(r[3], r[4], r[5], r[6]),
        nullable: r[7] === "1",
        identity: r[8] === "1",
        computed: r[9] === "1",
      },
      into: "columns",
    }),
  },
  {
    name: "indexes",
    sql: () => `SELECT s.name, o.name, i.name, i.is_primary_key, i.is_unique, c.name
FROM sys.indexes i
JOIN sys.objects o ON o.object_id = i.object_id
JOIN sys.schemas s ON s.schema_id = o.schema_id
JOIN sys.index_columns ic ON ic.object_id = i.object_id AND ic.index_id = i.index_id
JOIN sys.columns c ON c.object_id = ic.object_id AND c.column_id = ic.column_id
WHERE ${IN_SCOPE()} AND i.type > 0 AND ic.is_included_column = 0
ORDER BY s.name, o.name, i.index_id, ic.key_ordinal`,
    shape: (r) => ({
      owner: `${r[0]}.${r[1]}`,
      value: { name: r[2], primary: r[3] === "1", unique: r[4] === "1", column: r[5] },
      into: "indexRows",
    }),
  },
  {
    name: "foreignKeys",
    sql: () => `SELECT ps.name, po.name, pc.name, rs.name, ro.name, rc.name
FROM sys.foreign_key_columns fkc
JOIN sys.objects po ON po.object_id = fkc.parent_object_id
JOIN sys.schemas ps ON ps.schema_id = po.schema_id
JOIN sys.columns pc ON pc.object_id = fkc.parent_object_id AND pc.column_id = fkc.parent_column_id
JOIN sys.objects ro ON ro.object_id = fkc.referenced_object_id
JOIN sys.schemas rs ON rs.schema_id = ro.schema_id
JOIN sys.columns rc ON rc.object_id = fkc.referenced_object_id AND rc.column_id = fkc.referenced_column_id
ORDER BY ps.name, po.name, pc.name`,
    shape: (r) => ({
      owner: `${r[0]}.${r[1]}`,
      value: { column: r[2], references: `${r[3]}.${r[4]}.${r[5]}` },
      into: "foreignKeys",
    }),
  },
  {
    name: "parameters",
    sql: () => `SELECT s.name, o.name, p.name, t.name, p.max_length, p.is_output
FROM sys.parameters p
JOIN sys.objects o ON o.object_id = p.object_id
JOIN sys.schemas s ON s.schema_id = o.schema_id
JOIN sys.types t ON t.user_type_id = p.user_type_id
WHERE ${IN_SCOPE()}
ORDER BY s.name, o.name, p.parameter_id`,
    shape: (r) => ({
      owner: `${r[0]}.${r[1]}`,
      value: { name: r[2] || "(returns)", type: `${r[3]}${widthOf(r[3], r[4], 0, 0)}`, output: r[5] === "1" },
      into: "parameters",
    }),
  },
  {
    name: "synonyms",
    sql: () => `SELECT s.name, sn.name, sn.base_object_name
FROM sys.synonyms sn
JOIN sys.schemas s ON s.schema_id = sn.schema_id
ORDER BY s.name, sn.name`,
    shape: (r) => ({ owner: `${r[0]}.${r[1]}`, value: r[2], into: "target" }),
  },
];

// varchar(50), decimal(18,2), int — the width a reader needs, not the raw
// catalog numbers. max_length is bytes, so nvarchar halves it and -1 is (max).
const SIZED = ["varchar", "nvarchar", "char", "nchar", "varbinary", "binary"];
const SCALED = ["decimal", "numeric"];

function widthOf(type, maxLength, precision, scale) {
  if (SCALED.includes(type)) return `(${precision},${scale})`;
  if (!SIZED.includes(type)) return "";
  if (Number(maxLength) === -1) return "(max)";
  const chars = type.startsWith("n") ? Number(maxLength) / 2 : Number(maxLength);
  return `(${chars})`;
}

// Folds shaped rows into a map of object key -> structure fields.
function collect(shaped) {
  const byOwner = new Map();
  for (const { owner, value, into } of shaped) {
    const entry = byOwner.get(owner) ?? {};
    if (into === "target") entry.target = value;
    else (entry[into] = entry[into] ?? []).push(value);
    byOwner.set(owner, entry);
  }
  return byOwner;
}

// Index rows arrive one row per column; fold them into one entry per index and
// lift the primary key out, since a reader wants it named separately.
function foldIndexes(indexRows = []) {
  const byName = new Map();
  for (const row of indexRows) {
    const entry = byName.get(row.name) ?? { name: row.name, primary: row.primary, unique: row.unique, columns: [] };
    entry.columns.push(row.column);
    byName.set(row.name, entry);
  }
  const all = [...byName.values()];
  return {
    primaryKey: all.find((i) => i.primary)?.columns ?? [],
    indexes: all.filter((i) => !i.primary).map(({ name, unique, columns }) => ({ name, unique, columns })),
  };
}

// Runs every source against one target, health-checked between each, and
// returns a map of "schema.name" -> structure ready for fingerprinting.
async function fetchStructure(target, deps = {}) {
  const run = deps.run ?? sqlcmd.run;
  const shaped = [];

  for (const source of SOURCES) {
    await health.assertHealthy(target, `structure/${source.name} on ${target.logical}`, deps);
    const rows = await run({ host: target.host, database: target.physical, query: source.sql() }, deps);
    for (const row of rows) if (row.length >= 3) shaped.push(source.shape(row.map((c) => c.trim())));
    deps.logger?.info?.(`structure: ${target.logical}/${source.name} — ${rows.length} rows`);
  }

  const collected = collect(shaped);
  const structure = {};
  for (const [key, entry] of collected) {
    const { indexRows, ...rest } = entry;
    structure[key] = { ...rest, ...foldIndexes(indexRows) };
  }
  return structure;
}

module.exports = { SOURCES, widthOf, collect, foldIndexes, fetchStructure };
